Fusing All Gather with NLP Concat heads for llama attention

nardoTT opened the pull request:
### Ticket
Link to Github Issue https://github.com/tenstorrent/tt-metal/issues/18480

### Problem description
In the current implementation of llama attention, all_gather and NLP concat run as separate ops

### What's changed
Fusing All_Gather and NLP_concat_heads:
- The all_gather_concat implementation works with 3 links for the llama attention spec
- The worker cores are either executing all_gather + concat or just NLP concat
- All_gather_concat overlaps the NLP concat operating on local data (1/4 of input data) with the all gather operation
- NLP concat waits for the all_gather to finish to start the execution of the remaining 3/4 input data
- We add a signal semaphore that is mcasted by the all gather drain core to the other worker cores to signal the end of the all gather op
- The intermediate data is stored in an intermediate tensor. All gather op is writing to that tensor, and NLP concat heads op is reading from it

Adding Unit tests for T3k/TG for all_gather_concat op

This cuts the kernel device time of the ops by 2.65 us (from 19.13 us to 16.48 us) and the op to op latency by half (from 1.37 us to 0.653 us) in llama model

Model integration will be added in another PR


### Checklist
- [x] TG frequent tests: https://github.com/tenstorrent/tt-metal/actions/runs/14314529200
- [x]  TG unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/14314522233
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

djordje-tt committed:
rebase

nardoTT committed:
all_gather concat_heads fusion

nardoTT committed:
change semaphore index

nardoTT committed:
llama model integration for all_gather_concat op

nardoTT committed:
overlap local nlp

nardoTT committed:
fix program cache indices

nardoTT committed:
hardcode some rt args

nardoTT committed:
clean code

nardoTT committed:
rebase and add ccl optimizations

nardoTT committed:
add buffer tensor from python

nardoTT committed:
update llama_ccl for trace

nardoTT committed:
remove decode changes

nardoTT committed:
remove header file

jvegaTT commented on models/demos/llama3_subdevices/tests/tg_perf_unit_tests/test_ccl_async_perf_TG_llama.py:
```diff
@@ -122,3 +122,53 @@ def test_ar_tg_llama_perf(
+    command = f"pytest tests/ttnn/unit_tests/operations/ccl/test_minimals.py::test_concat_fuse -k {ar_type}"
+    cols = ["DEVICE KERNEL"]
+    op_name = "AllGatherConcat"
+    warmup_iters = warmup_iters * 4  # 5 iterations per device
```
Should be *32 as this is targeting a galaxy

jvegaTT commented on tests/ttnn/unit_tests/operations/ccl/fusion_subtests/concat_fuse_test.py:
```diff
@@ -0,0 +1,595 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
2025

jvegaTT commented on tests/ttnn/unit_tests/operations/ccl/test_minimals.py:
```diff
@@ -0,0 +1,116 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
2025

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat.cpp:
```diff
@@ -0,0 +1,130 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
```
2025

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat.cpp:
```diff
@@ -0,0 +1,130 @@
+    auto shard_spec =
+        ShardSpec(CoreRangeSet(std::vector{core_range_1, core_range_2}), shard_shape, ShardOrientation::ROW_MAJOR);
+    auto inter_output_mem_config =
+        ttnn::MemoryConfig(TensorMemoryLayout::HEIGHT_SHARDED, ttnn::BufferType::L1, shard_spec);
```
I may be wrong but I think you should be able to delete lines 27 to 32, they don't seem to get used at all afterwards

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat.hpp:
```diff
@@ -0,0 +1,79 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
```
2025

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat_pybind.cpp:
```diff
@@ -0,0 +1,138 @@
+            py::arg("subdevice_id") = std::nullopt,
+            py::arg("enable_persistent_fabric_mode") = false,
+            py::arg("queue_id") = DefaultQueueId},
+
```
Same generality question, does the llama model need both invocation pathways? Less flexibility = less potential bugs

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat_pybind.cpp:
```diff
@@ -0,0 +1,138 @@
+            cluster_axis (int): Provided a MeshTensor, the axis corresponding to MeshDevice to perform the line-all-gather operation on.
+            mesh_device (MeshDevice): Device mesh to perform the line-all-gather operation on.
+        * cluster_axis and mesh_device parameters are applicable only for Linear Topology.
+        Mesh Tensor Programming Guide : https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/Programming%20Mesh%20of%20Devices/Programming%20Mesh%20of%20Devices%20with%20TT-NN.md
```
This description doesn't seem to match your invoke

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat_pybind.cpp:
```diff
@@ -0,0 +1,138 @@
+            py::kw_only(),
+            py::arg("num_links") = 1,
+            py::arg("memory_config") = std::nullopt,
+            py::arg("topology") = ttnn::ccl::Topology::Ring,
```
Given ring is not supported in fabric this really shouldn't be the default here

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_op.cpp:
```diff
@@ -0,0 +1,344 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
```
2025

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_op.hpp:
```diff
@@ -0,0 +1,188 @@
+
+enum class AllGatherConcatVersion {
+    LLAMA_MINIMAL_SHARDED = 2,
+};
```
Do we really need a single element enum?

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_op.cpp:
```diff
@@ -0,0 +1,344 @@
+        output_shape, tt::tt_metal::TensorLayout(input_tensor.get_dtype(), tt::tt_metal::Layout::TILE, mem_config))};
+}
+
+AllGatherConcatVersion AllGatherConcat::select_version(const Tensor& input_tensor) const {
```
This is not a required function in op creation, it was only in all gather because the native one has multiple versions but here it just adds overhead

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,665 @@
+    auto reserved_packet_header_CB_handle =
+        CreateCircularBuffer(program, sender_worker_core_range, cb_reserved_packet_header_config);
+
+    uint32_t q_output_cb_index = tt::CBIndex::c_16;
```
You can get a decent perf increase by packing the CBs, that means for instance here if we have 3 CBs we use c_in0, c_in1, and c_in2

SeanNijjar replied:
I missed this. Please specify all CBs as CT args

SeanNijjar replied:
This is marked resolved but not addressed? (the diff is still showing c_16 in use)

nardoTT replied:
yeah we discussed it on slack, and after trying to pack the cbs, the perf didn't change. We assumed it is because c_16 is an output buffer for instance. And maybe packing helps when doing it for input cbs together or output cbs together. If you have any insight about it, that would be helpful

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/kernels/llama_all_gather_concat_reader.cpp:
```diff
@@ -0,0 +1,313 @@
+    // Load the input tensor spec
+    address_t tensor_address0 = get_arg_val<address_t>(arg_idx++);
+    uint32_t out_ready_sem_bank_addr_concat = get_arg_val<uint32_t>(arg_idx++);
+    uint32_t num_tiles_per_core = 4;  // get_arg_val<uint32_t>(arg_idx++);
```
make this a constexpr uint32_t

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/kernels/llama_all_gather_concat_reader.cpp:
```diff
@@ -0,0 +1,313 @@
+    DPRINT << "num_tiles_per_core: " << (uint32_t)num_tiles_per_core << "\n";
+    DPRINT << "num_tiles_to_read: " << (uint32_t)num_tiles_to_read << "\n";
+    DPRINT << "first_core_tile_start_offset: " << (uint32_t)first_core_tile_start_offset << "\n";
+    DPRINT << "num_cores: " << (uint32_t)num_cores << "\n";
```
Please delete all of these dprints

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/kernels/llama_all_gather_concat_writer.cpp:
```diff
@@ -0,0 +1,398 @@
+    DPRINT << "fabric_connection arg 1" << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
+    DPRINT << "fabric_connection arg 2" << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
+    DPRINT << "fabric_connection arg 3" << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
+    DPRINT << "fabric_connection arg 4" << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
```
Delete all the DPRINTs in the kernel

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/kernels/llama_all_gather_concat_writer.cpp:
```diff
@@ -0,0 +1,398 @@
+
+    // 3. wait for mcast output ready semaphore
+    if (wait_output_semaphore) {
+        while (*reinterpret_cast<volatile uint32_t*>(out_ready_sem_bank_addr) < out_ready_sem_wait_value);
```
Change < to != so it is more exact, less error prone

jvegaTT replied:
Or actually to make it cleaner use the noc_semaphore_wait api to do this

jvegaTT reviewed (commented):
I added some comments to the code, mostly small stuff. Please run APC as well as the tests Johanna mentioned. Also I don't see you integrating this with the model.

nardoTT committed:
post pr fix 1

nardoTT commented:
> I added some comments to the code, mostly small stuff. Please run APC as well as the tests Johanna mentioned. Also I don't see you integrating this with the model.

The perf values are from the model test, and the op is integrated in llama_ccl.py and llama_attention.py

jvegaTT reviewed (approved):
Sorry, not sure how I missed that. Looks really good!

nardoTT committed:
remove unused op infra

nardoTT committed:
remove version log

nardoTT committed:
undo model integration

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat.hpp:
```diff
@@ -0,0 +1,54 @@
+    static ttnn::Tensor invoke(
+        QueueId queue_id,
+        const ttnn::Tensor& input_tensor,
+        ttnn::Tensor& buffer_tensor,
```
Is this the buffer for the output of all-gather?

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_op.cpp:
```diff
@@ -0,0 +1,266 @@
+    auto batch = input_shape[1];
+    auto head_dim = input_shape[3];
+    // pad batch to 32 if necessary
+    if (batch < 32) {
```
why 32? Can this be defined in some constexpr variable instead of literal?

nardoTT replied:
32 is the batch size. This is specific for llama shape

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,665 @@
+    const size_t num_targets_backward =
+        line_topology.get_distance_to_end_of_line(ttnn::ccl::EdmLineFabricOpInterface::Direction::BACKWARD);
+
+    uint32_t batch_size = 1;
```
it would be good with all these hardcoded values to have a brief description of them. Additionally, with these hardcoded values - I assume there are a fair bit of limitations on what is supported? I didn't see any fatals to check for things like this. 

Can you please add some basic checks to the op that throw when any assumptions about op configuration are violated?

nardoTT replied:
I have added fatals for the shape and the core grid for the input

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,665 @@
+    const uint32_t num_cores = q_cores.num_cores();  // number of cores of the output
+    const auto& cores = corerange_to_cores(q_cores, num_cores, true);
+
+    auto core_range_1 = CoreRange({1, 1}, {3, 1});
```
It seems like more hardcoded assumptions? This is fine to have btw for these performance ops but it would be good to hoist the values and have them specified programatically (e.g. through constexpr variable or some op config struct) so that they are all organized in one place.

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,665 @@
+            .noc = reader_noc,
+            .compile_args = concat_reader_ct_args});
+
+    concat_reader_ct_args[6] = 2;
```
please name these literals.

nardoTT replied:
I have named this and other literals

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,665 @@
+        head_size,
+        batch,
+        head_tiles,
+        2,  // read the second phase
```
Please name -- in this case the comment could become the name.

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,665 @@
+
+    std::vector<uint32_t> nlp_local_core_x;
+    std::vector<uint32_t> nlp_local_core_y;
+    for (uint32_t k = 0; k < 8; k++) {
```
please name literal. Does this correspond to any of the previous literal values?

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,665 @@
+        const auto& core = cores[i];
+        std::vector<uint32_t> input_cores_x;
+        std::vector<uint32_t> input_cores_y;
+        for (uint32_t k = 0; k < 8; k++) {
```
same literal as earlier loop.

SeanNijjar reviewed (changes requested):
I am requesting changes due to the footguns. I think there are quite a few places that need safety checks to avoid something like slightly modifying the subdevice core grid to causing silent undefined behaviour in the op

nardoTT committed:
post pr fix 2

nardoTT committed:
fix hooks

nardoTT committed:
fix hooks 2

nardoTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/kernels/llama_all_gather_concat_reader.cpp:
```diff
@@ -0,0 +1,283 @@
+        }
+    };
+
+    auto nlp_concat = [&](uint32_t head_size_num_tiles,
```
I moved the CTs outside of the main function and have changed the lambdas to functions

sraizada-tt commented:
can you please run TG CI? unit + frequent

avoraTT commented on models/demos/llama3_subdevices/tests/tg_perf_unit_tests/test_ccl_async_perf_TG_llama.py:
```diff
@@ -172,3 +172,53 @@ def test_rms_perf(
+
+
+@pytest.mark.parametrize(
+    "ar_type, warmup_iters, perf_target_us",
```
can we get rid of `ar_type` since this is not an all reduce? also since there's only one type, we probably don't need it at all

nardoTT replied:
sure, I removed it

nardoTT committed:
remove ar type from perf test

tt-aho commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/kernels/llama_all_gather_concat_writer.cpp:
```diff
@@ -0,0 +1,338 @@
+        in_tile_offset_by_head);
+
+    if (reset_global_semaphore) {
+        const uint64_t dest_noc_addr = get_noc_addr(my_x[0], my_y[0], out_ready_sem_bank_addr);
+        noc_inline_dw_write(dest_noc_addr, 0);
```
You don't need a noc write here. This should just be a local risc write, ex
`*reinterpret_cast<volatile tt_l1_ptr uint32_t*>(out_ready_sem_bank_addr) = 0;`

nardoTT replied:
you are right, I changed it

nardoTT commented:
> can you please run TG CI? unit + frequent

done

nardoTT committed:
remove unnecessary noc write

nardoTT committed:
modify perf test

nardoTT commented on tests/ttnn/unit_tests/operations/ccl/test_minimals.py:
```diff
@@ -411,3 +416,96 @@ def test_tg_trace_rms_fuse(
+        # ttnn.bfloat8_b,
+    ],
+)
+@pytest.mark.parametrize("num_iters, warmup_iters", [[30, 5]])
```
done

nardoTT commented on tests/ttnn/unit_tests/operations/ccl/fusion_subtests/concat_fuse_test.py:
```diff
@@ -0,0 +1,603 @@
+        logger.info(f"Using Async Mode for All Gather Op Dispatch")
+
+    compute_grid_size = mesh_device.compute_with_storage_grid_size()
+    ccl_sub_device_crs = ttnn.CoreRangeSet(
```
this part is actually for the fake fused test, it is not being called. I deleted this section

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_op.cpp:
```diff
@@ -0,0 +1,277 @@
+        input_tensor.memory_config().memory_layout);
+
+    TT_FATAL(
+        input_core_ranges[0].start_coord.x == 1 && input_core_ranges[0].end_coord.x == 3 &&
```
Is there a fatal for size to ensure the second half of this fatal isn't OoB?

nardoTT replied:
I have added a FATAL to check grid bounds

SeanNijjar commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/kernels/llama_concat_reader.cpp:
```diff
@@ -0,0 +1,224 @@
+    uint32_t arg_idx = 3 + 2 * in_num_cores;
+    uint32_t tensor_address0 = get_arg_val<uint32_t>(arg_idx);
+
+    std::array<uint32_t, 8> core_noc_x = {19, 20, 21, 19, 20, 21, 19, 20};
```
I also don't see this resolved and asserted against in host program?

SeanNijjar reviewed (commented):
There are (or appear to be) multiple unresolved comments marked as resolved when I am reviewing this again

nardoTT commented:
> There are (or appear to be) multiple unresolved comments marked as resolved when I am reviewing this again

for the assertion of the cores (core_noc_x and out_ready_sem_noc0_x_concat and out_ready_sem_noc0_y_concat), I added the FATAL for the input core grid, cause these values reflect the location of the input tensors. So, the test will FATAL out if we can't use these values. I am not sure if the host code assertion would be different

nardoTT committed:
delete commented lines

nardoTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,693 @@
+        for (auto nocy : noc_y_coords) {
+            reader_rt_args.push_back(nocy);
+        }
+        // reader_rt_args.push_back(ring_size * num_links);  // sem target value
```
done

nardoTT commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/device/all_gather_concat_program.cpp:
```diff
@@ -0,0 +1,693 @@
+            output_tensor_cores_x.size(),         // num_cores
+            wait_output_semaphore,                // wait_output_semaphore
+            reset_global_semaphore,               // reset_global_semaphore
+            // drain_sync_core.x,                     // out_ready_sem_noc0_x
```
done

SeanNijjar commented:
> > There are (or appear to be) multiple unresolved comments marked as resolved when I am reviewing this again
> 
> for the assertion of the cores (core_noc_x and out_ready_sem_noc0_x_concat and out_ready_sem_noc0_y_concat), I added the FATAL for the input core grid, cause these values reflect the location of the input tensors. So, the test will FATAL out if we can't use these values. I am not sure if the host code assertion would be different

Specifically for this comment https://github.com/tenstorrent/tt-metal/pull/19788/files#r2029228705 and another very similar one I don't see the corresponding checks. Since the coords are hardcoded in the kernel, there should also be a check in the program factory that the logical to virtual coord mapping is actually as you expect (for example, it may be different in BH and if that's the case, your changes should fail explicitly in the program factory, in case it's accidentally used in BH -- alternatively, throw that we aren't running this op on BH).

nardoTT committed:
add asserts for hardcoded values

nardoTT commented:
> > > There are (or appear to be) multiple unresolved comments marked as resolved when I am reviewing this again
> > 
> > 
> > for the assertion of the cores (core_noc_x and out_ready_sem_noc0_x_concat and out_ready_sem_noc0_y_concat), I added the FATAL for the input core grid, cause these values reflect the location of the input tensors. So, the test will FATAL out if we can't use these values. I am not sure if the host code assertion would be different
> 
> Specifically for this comment https://github.com/tenstorrent/tt-metal/pull/19788/files#r2029228705 and another very similar one I don't see the corresponding checks. Since the coords are hardcoded in the kernel, there should also be a check in the program factory that the logical to virtual coord mapping is actually as you expect (for example, it may be different in BH and if that's the case, your changes should fail explicitly in the program factory, in case it's accidentally used in BH -- alternatively, throw that we aren't running this op on BH).
> 
> Additionally, maybe the coords change in the host program factory but accidentally get missed in the kernel and the fatals would ensure they're updated in the kernels too (it's for this reason it would be better to just fatal directly on the translated coords themselves rather than arch)

yeah that makes sense, thank you for pointing it out. I have added asserts in the host program for the hardcoded values

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat.hpp:
```diff
@@ -0,0 +1,54 @@
+        std::optional<tt::tt_metal::SubDeviceId> subdevice_id = std::nullopt,
+        bool enable_persistent_fabric_mode = false);
+
+    static ttnn::Tensor invoke(
```
you can remove this version. given a version where queueId is provided as a first argument, infra will make sure to support calls with both cqid provided and not.

nardoTT replied:
done

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_concat_heads_fused/all_gather_concat.cpp:
```diff
@@ -0,0 +1,71 @@
+        enable_persistent_fabric_mode);
+}
+
+ttnn::Tensor ExecuteAllGatherConcat::invoke(
```
can be removed. left a detailed info in a similar comment in .hpp

nardoTT replied:
done

nardoTT committed:
remove invoke without qid

nardoTT committed:
change header file
